### PR TITLE
StreamBatchSource and TPFSSink

### PR DESCRIPTION
**Pull request by rohitsinha54**
Initial implementation of StreamBatchSource and TPFSSink. 
Also added an integration test which tests the pipeline below:

StreamBatchSource -> StreamEventToStructuredRecordTransform -> StructuredRecordToAvroTransform -> TPFS

Issues: 
https://issues.cask.co/browse/CDAP-1945
https://issues.cask.co/browse/CDAP-1947

### CODE REVIEW

**Comment by gokulavasan** [comment]
@rohitsinha54 Can you start a bamboo run under the Adapters config?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/pom.xml [hunk 1]
```diff
@@ -66,6 +66,17 @@
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.avro</groupId>
+      <artifactId>avro</artifactId>
+      <version>1.7.7</version>
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/pom.xml]: version should come from the parent pom

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+/**
+ * Tests for {@link ETLBatchTemplate} for Stream conversion from stream to avro format for writing to
+ * {@link TimePartitionedFileSet}
+ */
+public class ETLStreamConversionTest extends TestBase {
+
+  private static final Gson GSON = new Gson();
+
+  Schema bodySchema = Schema.recordOf(
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java]: these schemas can be private static final as well

File: cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml [hunk 3]
```diff
@@ -36,6 +36,99 @@
...
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.11</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.avro</groupId>
+      <artifactId>avro</artifactId>
+      <version>1.7.6</version>
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml]: versions should come from parent. (notice how the other pom has 1.7.7, which is why we want it in parent)

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.avro.mapreduce.AvroJob;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.Job;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link BatchSink} to write to {@link TimePartitionedFileSet}
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: to write Avro records to

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public class TimePartitionedFileSetDatasetAvroSink extends BatchSink<AvroKey<GenericRecord>, NullWritable> {
+
+  static final String DATASETNAME = "name";
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(TimePartitionedFileSetDatasetAvroSink.class.getName());
+    configurer.setDescription("An Avro sink for TimePartitionedFileSetDataset");
+    configurer.addProperty(new Property("schema", "The schema of the record", true));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: lets be consistent and have a constant for this property name as well

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  public void prepareJob(BatchSinkContext context) {
+    Map<String, String> sinkArgs = Maps.newHashMap();
+    TimePartitionedFileSetArguments.setOutputPartitionTime(sinkArgs, context.getLogicalStartTime());
+    TimePartitionedFileSet sink = context.getDataset(context.getRuntimeArguments().get(DATASETNAME), sinkArgs);
+    context.setOutput(context.getRuntimeArguments().get(DATASETNAME), sink);
+    Schema userSchema = null;
+    try {
+      userSchema = Schema.parseJson(context.getRuntimeArguments().get("schema"));
+    } catch (IOException e) {
+      Throwables.propagate(e);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: would be good to log a messag

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: actually, there is no need to use CDAP Schema class at all. just use avro's schema parser directly on the property value

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+   * Prepare the Batch Job. Used to configure the Hadoop Job before starting the Batch Job.
+   *
+   * @param context {@link BatchSourceContext}
+   */
+  @Override
+  public void prepareJob(BatchSourceContext context) {
+    long endTime = context.getLogicalStartTime();
+    //TODO: Once the method to get the frequency from the schedule is added change it to use that. Then we will not
+    // need the frequency as a configuration here
+    long startTime = endTime - Long.valueOf(context.getRuntimeArguments().get("frequency")) * 1000L;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: we shouldn't use milliseconds as the unit.  The smallest unit possible in a schedule is a minute, so lets use minutes.

**Comment by rohitsinha54** [comment]
This PR has some transform commits which are already merged. I am closing this for now and will open again.

**Comment by rohitsinha54** [comment]
Build running here: http://builds.cask.co/browse/CDAP-DT6-1

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java [hunk 8]
```diff
@@ -41,4 +46,28 @@ public void configure(StageConfigurer configurer) {
    * @param context {@link BatchSourceContext}
    */
   public abstract void prepareJob(BatchSourceContext context);
+
+  protected long parseFrequency(String frequencyStr) {
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java]: Gokul: This method is copied from ConversionConfig which is in cdap-adapter modules. That module is going to go away right ?

File: pom.xml [hunk 9]
```diff
@@ -142,6 +142,7 @@
   <modules>
     <module>cdap-common-unit-test</module>
     <module>cdap-api</module>
+    <module>cdap-app-templates</module>
```
Comment (gokulavasan) [inline pom.xml]: Don't need to add cdap-app-templates. It is built with when templates profile is enabled.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java (hunk 8 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java]: that module will go away

File: cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml [hunk 10]
```diff
@@ -35,6 +35,11 @@
       <groupId>co.cask.cdap</groupId>
       <artifactId>cdap-api</artifactId>
     </dependency>
+    <dependency>
+      <groupId>commons-lang</groupId>
+      <artifactId>commons-lang</artifactId>
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml]: lets avoid pulling in this dependency

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java [hunk 11]
```diff
@@ -41,4 +46,28 @@ public void configure(StageConfigurer configurer) {
...
+    //TODO: replace with TimeMathParser (available only internal to cdap)
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(frequencyStr));
+    frequencyStr = frequencyStr.toLowerCase();
+
+    String value = frequencyStr.substring(0, frequencyStr.length() - 1);
+    Preconditions.checkArgument(StringUtils.isNumeric(value));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java]: lets do this without pulling in commons lang

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java (hunk 8 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java]: this shouldn't be in BatchSource as it's not related to source. Can pull it out to some common helper class.

**Comment by albertshau** [comment]
couple minor comments

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    long endTime = context.getLogicalStartTime();
+    //TODO: Once the method to get the frequency from the schedule is added change it to use that. Then we will not
+    // need the frequency as a configuration here (depends on to CDAP-2048)
+    long startTime = endTime - parseFrequency(context.getRuntimeArguments().get("frequency"));
+
+    String streamName = context.getRuntimeArguments().get("streamName");
+    LOG.info("Setting input to Stream : {}", streamName);
+    Schema schema = Schema.recordOf("streamEvent", Schema.Field.of("body", Schema.of(Schema.Type.STRING)));
+
+    // TODO: This is not clean.
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: Can you file a JIRA for this?

File: cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml (hunk 10 above)
Comment (henry-cask) [inline cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml]: Why avoid common-lang into this module? Is the idea to keep api as slim as possible?

Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml]: Yes. api should have no dependency except cdap-api. It simplifies classloading.

Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml]: Because classes in api module are for plugin to implement. The more dependencies it has, the higher chance there will be conflict in library version/classes.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    context.setInput(new StreamBatchReadable(streamName, startTime, endTime,
+                                             "co.cask.cdap.data.stream.decoder.IdentityStreamEventDecoder")
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: Why? The MapReduce runtime is smart enough to inspect the Mapper type and be able to choose the right decoder if the mapper is `extends Mapper<LongWritable, StreamEvent>`.

Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: This class name is referring to a cdap internal, which should never be in app-template.

File: cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java [hunk 14]
```diff
@@ -223,9 +223,22 @@ public StreamBatchReadable(String streamName, long startTime, long endTime) {
...
+  /**
+   * Constructs an instance with the given properties.
+   *
+   * @param streamName Name of the stream
+   * @param startTime Start timestamp in milliseconds (inclusive) of stream events provided to the job
+   * @param endTime End timestamp in milliseconds (exclusive) of stream events provided to the job
+   * @param decoderTypeClassname The full qualified name of the class for decoding {@link StreamEvent}
+   */
+  public StreamBatchReadable(String streamName, long startTime,
+                             long endTime, String decoderTypeClassname) {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/data/stream/StreamBatchReadable.java]: Shouldn't need this method. See my comment below about not to use this method to set a CDAP internal class name.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java [hunk 15]
```diff
@@ -41,4 +46,28 @@ public void configure(StageConfigurer configurer) {
...
+    Integer parsedValue = Integer.valueOf(value);
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSource.java]: Should use `Integer.parseInt`. That will through NumberFormatException if it is not a number.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/pom.xml [hunk 16]
```diff
@@ -66,6 +66,17 @@
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.avro</groupId>
+      <artifactId>avro</artifactId>
+      <version>${apache.avro.version}</version>
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/pom.xml]: Why specify version here? We should be using dependency management in the parent pom.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+import co.cask.cdap.templates.etl.common.config.ETLStage;
+import co.cask.cdap.templates.etl.transforms.GenericTypeToAvroKeyTransform;
+import co.cask.cdap.templates.etl.transforms.StreamToStructuredRecordTransform;
+import co.cask.cdap.templates.etl.transforms.StructuredRecordToGenericRecordTransform;
+import co.cask.cdap.test.ApplicationManager;
+import co.cask.cdap.test.DataSetManager;
+import co.cask.cdap.test.MapReduceManager;
+import co.cask.cdap.test.StreamWriter;
+import co.cask.cdap.test.TestBase;
+import com.clearspring.analytics.util.Lists;
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java]: Should use guava Lists.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * Tests for {@link ETLBatchTemplate} for Stream conversion from stream to avro format for writing to
+ * {@link TimePartitionedFileSet}
+ */
+public class ETLStreamConversionTest extends TestBase {
+
+  private static final Gson GSON = new Gson();
+
+  private static final Schema bodySchema = Schema.recordOf(
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java]: static final variable should be `ALL_CAPS`.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+    "event",
+    Schema.Field.of("ticker", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("num", Schema.of(Schema.Type.INT)),
+    Schema.Field.of("price", Schema.of(Schema.Type.DOUBLE)));
+
+  private static final Schema eventSchema = Schema.recordOf(
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java]: static final variable should be `ALL_CAPS`.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml [hunk 20]
```diff
@@ -117,12 +117,17 @@
     <dependency>
       <groupId>org.apache.avro</groupId>
       <artifactId>avro</artifactId>
-      <version>1.7.6</version>
+      <version>${apache.avro.version}</version>
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml]: Should be using dependency management in the parent pom, meaning no version should be specified here.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.hadoop.mapreduce.Job;
+
+import java.util.Map;
+
+/**
+ * A {@link BatchSink} to write Avro record to {@link TimePartitionedFileSet}
+ */
+public class TimePartitionedFileSetDatasetAvroSink extends BatchSink<AvroKey<GenericRecord>, NullWritable> {
+
+  static final String DATASETNAME = "name";
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Why not private?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+  public static final String SCHEMA = "schema";
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Why public?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java (hunk 21 above)
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: And the name should be `DATASET_NAME`.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+import co.cask.cdap.templates.etl.api.batch.BatchSource;
+import co.cask.cdap.templates.etl.api.batch.BatchSourceContext;
+import org.apache.hadoop.io.LongWritable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link BatchSource} for {@link Stream} to use {@link Stream} as Source.
+ */
+public class  StreamBatchSource extends BatchSource<LongWritable, StreamEvent> {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: Remove extra space between `class` and `StreamBatchSource`.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java (hunk 13 above)
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: Or, in the ETL case, since the type is not available, we can modify MapReduceRuntimeService to be able to detect the StreamEventDecoder through a property set in the `Job`. E.g. we can introduce a property call `cdap.mapreduce.map.input.value.class` then can be set to indicate the class name of the Mapper input value type.

Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: It that case, it's not specific to template and is usable by any MapReduce program.

Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: As discussed in person we will do this in an independent PR.

Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: I would say you do that first before merging this. That should be a simple change.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * An Utility class for ETL
+ */
+public class ETLUtils {
+
+  /**
+   * Parses a frequency String to its long value
+   *
+   * @param frequencyStr the frequency string (ex: 5m, 5h etc).
+   * @return long equaivalent of frequency string
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java]: nit: typo: equivalent.  Also, should be more descriptive, something like "milliseconds equivalent of the frequency string"

File: cdap-app-templates/cdap-etl/pom.xml [hunk 25]
```diff
@@ -179,6 +180,18 @@
         <artifactId>javax.ws.rs-api</artifactId>
         <version>${rs-api.version}</version>
       </dependency>
+
+      <dependency>
+        <groupId>org.apache.avro</groupId>
+        <artifactId>avro</artifactId>
+        <version>1.7.7</version>
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/pom.xml]: ```${apache.avro.version}```

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+  private static final String SCHEMA = "schema";
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(TimePartitionedFileSetDatasetAvroSink.class.getSimpleName());
+    configurer.setDescription("An Avro sink for TimePartitionedFileSetDataset");
+    configurer.addProperty(new Property(SCHEMA, "The schema of the record", true));
+    configurer.addProperty(new Property(DATASET_NAME, "Name of the Time Partitioned FileSet Dataset to which the " +
+      "records have to be written", true));
+  }
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: after you merge develop, should create the dataset in configurePipeline()

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+public class StreamBatchSource extends BatchSource<LongWritable, StreamEvent> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamBatchSource.class);
+
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(StreamBatchSource.class.getSimpleName());
+    configurer.setDescription("Use Stream as Source");
+    configurer.addProperty(new Property("streamName", "Name of the stream to use as Source", true));
+    configurer.addProperty(new Property("frequency", "Frequency of the schedule", false));
+  }
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: after you merge develop, should add the stream in configurePipeline()

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An Utility class for ETL
+ */
+public class ETLUtils {
+
+  /**
+   * Parses a frequency String to its long value
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java]: More details. Want somebody to look at javadoc and be able to understand what it does. Frequency is a number followed by a unit, with 's' for seconds, 'm' for minutes, 'h' for hours, and 'd' for days.  For example, '5m' means 5 minutes.

**Comment by albertshau** [comment]
couple minor comments after you merge the related PR

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java (hunk 27 above)
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/StreamBatchSource.java]: Yes working on that for both source and sink

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+public class TimePartitionedFileSetDatasetAvroSink extends BatchSink<AvroKey<GenericRecord>, NullWritable> {
+
+  private static final String TPFS_NAME = "name";
+  private static final String SCHEMA = "schema";
+  private static final String BASE_PATH = "basePath";
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(TimePartitionedFileSetDatasetAvroSink.class.getSimpleName());
+    configurer.setDescription("An Avro sink for TimePartitionedFileSetDataset");
```
Comment (nitinmotgi) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: "AVRO Sink with Time Partitioned File Dataset"

**Comment by albertshau** [comment]
:+1:

**Comment by rohitsinha54** [comment]
Final build running here: http://builds.cask.co/browse/CDAP-DT6-3

**Comment by rohitsinha54** [comment]
The build fails because of unit test failures due to recent changes on develop here: 
https://github.com/caskdata/cdap/blob/develop/cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/templates/etl/batch/ETLBatchTemplate.java#L161

Which doesn't seem like doing the right check. Me and @gokulavasan are working to fix the check.

**Comment by rohitsinha54** [comment]
Updated with develop and new build running here: 
http://builds.cask.co/browse/CDAP-DT6-4

**Comment by rohitsinha54** [comment]
New Build: http://builds.cask.co/browse/CDAP-DT6-5
